### PR TITLE
Core/Garrison Major Improvement

### PR DESCRIPTION
**Pull request by Traesh**
**Changes proposed:**

- Change garrison code organisation, allow player to have multiple garrisons from different types (WodGarrison, ClassHall, and soon(tm) WarCampaign for BFA)
- Create new "WodGarrison" and "ClassHall" classes, inherit from "Garrison" base class
- Add "GarrisonAI" script class, DB assigned
- Handle player in/out of garrison by walking in right zone
- Add 2 missing opcodes & fix 2 others
- Handle 17 garrison opcodes (Garrison upgrade, ...)
- Load 4 new DB2 (GarrFollowerLevelXP, GarrFollowerQuality, GarrFollowerType, GarrMission)
- Handle major part of mission systems (missing chance calculation & reward selection)
- Fix Alliance building validation in wod garrisons

Minor additions :
- Add player "SeamlessTeleportToMap" helper
- Add player "AddMovieDelayedTeleport" helper to teleport a player after a movie completion
- Fix Garrison Hearthstone

**Target branch(es):**

- [x] master

**Issues addressed:** Closes #  (insert issue tracker number)
none ATM

**Tests performed:** (Does it build, tested in-game, etc.)
Build, tested in-game

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] Todo : mission success chance calculation (Will require to copy client code)
- [ ] Todo : mission rewards selection (wod rewards can be taken from DB2 deleted at Legion)

/!\ Attention /!\

Garrisons still need a massive work to be Blizzlike, and this is not this PR goal, see it as a major improvement, do not expect me to fully debug garrison before merging it

### CODE REVIEW

**Comment by Shauren** [comment]
fyi, entering/leaving is not areaid based on retail but coord based (i smell a serverside areatrigger)

File: src/server/game/Server/Protocol/Opcodes.h [hunk 1]
```diff
@@ -298,6 +298,7 @@ enum OpcodeClient : uint16
     CMSG_GARRISON_REQUEST_LANDING_PAGE_SHIPMENT_INFO  = 0x32DF,
     CMSG_GARRISON_REQUEST_SHIPMENT_INFO               = 0x32DE,
     CMSG_GARRISON_RESEARCH_TALENT                     = 0x32D1,
+    CMSG_GARRISON_REQUEST_SCOUTING_MAP                = 0x323B,
```
[the remark this answers is not in the document]
Comment (Traesh) [reply src/server/game/Server/Protocol/Opcodes.h]: It send back SMSG_GARRISON_SCOUTING_MAP_RESULT :trollface: 

This is the opcode sent when player right-click mission npc in class halls, result open the missions/followers/scoutingmap UI

File: src/server/game/Handlers/MovementHandler.cpp [hunk 2]
```diff
@@ -139,8 +139,7 @@ void WorldSession::HandleMoveWorldportAck()
     else
     {
         GetPlayer()->UpdateVisibilityForPlayer();
-        if (Garrison* garrison = GetPlayer()->GetGarrison())
-            garrison->SendRemoteInfo();
+        GetPlayer()->SendGarrisonRemoteInfo();
```
[the remark this answers is not in the document]
Comment (Traesh) [reply src/server/game/Handlers/MovementHandler.cpp]: Yes, and one packet contain all garrisons remote infos (WodGarrison + Class Halls (+ War Campaign soon)) that's why i moved it out of Garrison (which represent only one of those)

I can put this in GarrisonMgr with player as parameter if you prefer

File: src/server/game/Spells/SpellEffects.cpp [hunk 3]
```diff
@@ -5750,7 +5751,7 @@ void Spell::EffectGiveArtifactPowerNoBonus(SpellEffIndex /*effIndex*/)
     if (effectHandleMode != SPELL_EFFECT_HANDLE_LAUNCH_TARGET)
         return;
 
-    if (!unitTarget || unitTarget->GetTypeId() != TYPEID_PLAYER)
+    if (!unitTarget || !unitTarget->IsPlayer())
```
[the remark this answers is not in the document]
Comment (Traesh) [reply src/server/game/Spells/SpellEffects.cpp]: Well refactor-only PRs are refused so i was hoping this refactor would sneak away :D

[the remark this answers is not in the document]
Comment (joschiwald) [reply src/server/game/Spells/SpellEffects.cpp]: we could add some kind of macros here and drop all these different if statements and ToPlayer() calls in SpellEffects.cpp:

```c++
#define GET_PLAYER_TARGET_OR_RETURN(playerTarget) \
    if (!unitTarget) \
        return; \
    Player* playerTarget = unitTarget->ToPlayer(); \
    if (!playerTarget) \
        return

void Spell::EffectExample(SpellEffIndex /*effIndex*/)
{
    if (effectHandleMode != SPELL_EFFECT_HANDLE_HIT_TARGET)
        return;

    GET_PLAYER_TARGET_OR_RETURN(playerTarget);

    playerTarget->DoSomething();
}
```

File: src/server/game/Server/Protocol/Opcodes.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shauren) [reply src/server/game/Server/Protocol/Opcodes.h]: fyi, named them as ADVENTURE_MAP_POI_QUERY in 8.0.1

**Comment by Nyr97** [comment]
Any update on this report or was it abandoned?

**Comment by Traesh** [comment]
Well this PR was let to rot for 4 years without reviews so 🤷‍♂️

**Comment by MaxtorCoder** [comment]
> Well this PR was let to rot for 4 years without reviews so 🤷‍♂️

Well I mean, you did not push any commits in the last 4 years. :) Even after the last review.

**Comment by Traesh** [comment]
Which review ? Apart for the opcode name, the others are questions or proposals

**Comment by MaxtorCoder** [comment]
> Which review ? Apart for the opcode name, the others are questions or proposals

Still, it has been abandoned and the code structure is not good either, there is no need to split the Garrison classes from expansion. (Shauren's words on Discord)

**Comment by Traesh** [comment]
When there is no review and the code is never merged i don't know what i can possibly do about that.

And i can't guess Shauren's word on Discord, that's why there is a review system on Github. I still think we need to split the code since there are some big differences between wod garrison and the rest of them (Buildings). Of course now that we have 2 more expensions of feedback we can use that and just make the Wod garrison a specialized "BuildingGarrison" while the other can remain in the main "Garrison" class since everything is shared for all other expansions.

Going to close this one, after 4 years i clearly don't see myself fixing all the conflicts.

**Pull request closed without merging**